### ISSUE TITLE
Allow different Audio Player (i.e. Tone.js) to work with Peaks.js Waveform

### ISSUE THREAD

**Issue by tscz** [bbc/peaks.js#296]
Hello @chrisn,

this is more a feature request than an issue: It would be great if the peaks.js waveform could be linked with another player/library like Tone.js or Howler.js.

Currently the peaks.js waveform is tightly coupled with the <audio>-webaudio player. Out-of-the-box it is not possible to link the waveform with an external player (like Tone.js or Howler.js).

However, it is possible to replace the player with a compatible class but it comes with a price of code duplication, unofficial patching and using internal api (peaks.emit()). Very fragile, see https://github.com/tscz/transcribe/blob/ae85033b5edf768e60123e2b90cbf2894b36e4a3/src/features/audio/audioPlayer.ts#L20.

Is there any interest supporting this feature? If so, I would love to contribute here. 

Thx and regards
Teje

**Comment by tscz** [comment]
FYI: My usecase is that I want to use Peaks.js as the Waveform Viewer and Tone.js as a player for adding a metronome or drumbeat to the playback and changing pitch and playbackrate.

**Comment by chrisn** [comment]
I'm open to the idea of supporting Web Audio generated audio, as an option in addition to binding to a media element. I'd want to avoid taking a dependency on any specific Web Audio library though, so is there a way to do this using purely `AudioContext`? And I'd also want to avoid any breaking changes to the Peaks.js API for existing users.

**Comment by tscz** [comment]
I added a design sketch in https://github.com/tscz/peaks.js/commit/2f7bdf0c9ec30223885cdaf4b0bd7e5f224f3f9c.

The implentation idea is to split the Player in a general part, which can be extended by an adapter pattern and a default implementation based on html5 media element. Replacing the default player then boils down to define an adapter.
```javascript
var adapter = {
    isReady: function() {},
    destroy: function() {},
    play: function() {},
    pause: function() {},
    isPlaying: function() {},
    isSeeking: function() {},
    getCurrentTime: function() {},
    getDuration: function() {},
    seek: function(time) {},
    playSegment: function(segment) {}
}

CustomPlayer = new Player(peaks,adapter);
```
No dependency to another library, not even a dependency to the AudioContext. And also no breaking changes to the Peaks.js API for users. Is this going into the right direction for you?

**Comment by alittlebitweird** [comment]
I'm also very interested in this!

**Comment by chrisn** [comment]
@tscz Sorry for taking so long to reply. Yes, your proposal looks good to me. We would also need an option to set the adapter on initialisation:

```js
const playerAdapter = {
  play: () => { ... }
  // etc.
};

const options = {
  // etc.
  player: playerAdapter
};

Peaks.init(options, (err, peaks) => {
  // ...
});
```
So users have to pass either a  `mediaElement` option, to use the default `HTMLMediaElement` adapter, or a `player` option with a suitable adapter object.

**Comment by tscz** [comment]
@chrisn Cool, I'll have a look into it. One question about validation of the new configuration option:
* I guess a runtime validation for the options.player should be done.  Where and how should this be done? I see two ways:
** First option: In  Peaks.prototype._setOptions(), since it is the place for validating options. However, then there is a risk of forgetting the validation checks in options, if the abstract Player is changed somewhere in the future.
** Second option: Validation within the constructor of Player(). Comparable approach as in the Segment() constructor and the option.segments
I would vote for the second option, a central place for Player logic and adapter requirements/validation.

For typescript users this adapter interface would be part of the .d.ts type description file, so they would get compile time validation checks for the new option.

**Comment by chrisn** [comment]
The second way sounds good to me. Thanks!

**Comment by chrisn** [comment]
Resolved via #314.
